### ISSUE TITLE
[FEATURE] Support `databricks_grant`, `databricks_permission` etc. resources

### ISSUE THREAD

**Issue by karlschriek** [databricks/terraform-provider-databricks#1976]
### Use-cases

The databricks provider currently offers resources such as  `databricks_grants` and `databricks_permissions` (and probably several others that work in the same manner, but those are the two I am working with right now). These resource require you to assign _all_ grants/permissions for a single entity within the same resource.

For example (from the official docs):

```
resource "databricks_grants" "sandbox" {
  metastore = databricks_metastore.this.id
  grant {
    principal  = "Data Engineers"
    privileges = ["CREATE_CATALOG", "CREATE_EXTERNAL_LOCATION"]
  }
  grant {
    principal  = "Data Sharer"
    privileges = ["CREATE_RECIPIENT", "CREATE_SHARE"]
  }
}
```

The above is the _only_ place where I can add grants to the `databricks_metastore.this.id` resource. Any additional combinations of entity and privilege can only be added here. This makes composition in complicated terraform projects extremely difficult. We generally follow the principle of modular composition where each individual module is self-contained and additional, self-contained modules can be added without needing to change existing ones. With a resource such as `databricks_grants` this is not possible. What we would prefer to have would be to set a singular `databricks_grant` (and `databricks_permission` etc.) for each specific combination of (for example) metastore/principal/privileges.

This would be in line with how role assignments are for example dealt with by the Azure Resource Manager.

```terraform
resource "azurerm_role_assignment" "example2" {
  name                 = "role-assignement-1" //this field is optional. if left out a GUID will be generated
  scope                = var.my_resource_id
  principal_id         = var_principal_id
  role_definition_id   = var_role_definition_1_resource_id
}

resource "azurerm_role_assignment" "example2" {
  name                 = "role-assignement-2"
  scope                = var.my_resource_id
  principal_id         = var_principal_id
  role_definition_id   = var_role_definition_2_resource_id
}
```

This is possible because the "name" field is what uniquely defines the role assignment resource.

### Proposal

I would suggest allow for a `databricks_grant `(and similar for other resources) as follows:

```terraform
resource "databricks_grants" "data_scientists" {
  name = "data-scientists-1" // make this field optional; generate GUID if not set. This is the resource ID
  catalog = databricks_catalog.sandbox.name
  principal  = "Data Scientists"
  privileges = ["USE_CATALOG", "USE_SCHEMA", "CREATE_SCHEMA", "CREATE_TABLE", "MODIFY"]
}
resource "databricks_grants" "data_engineers" {
  name = "data-scientists-1" // make this field optional; generate GUID if not set. This is the resource ID
  catalog = databricks_catalog.sandbox.name
  principal  ="Data Engineers"
   privileges = ["USE_CATALOG", "USE_SCHEMA", "CREATE_SCHEMA", "CREATE_TABLE", "MODIFY"] 
}
```

**Comment by nkvuong** [comment]
@karlschriek 
we will need changes to the existing permissions/grants API for this - at the moment the API does not have the concept of role assignments and therefore drift configuration in Terraform will be very flaky

**Comment by karlschriek** [comment]
That's a shame, should I log a request somewhere for the API to be expanded?

**Comment by nkvuong** [comment]
@karlschriek if you mention it to your account team, I have also flagged it with the product team to put it on their backlog

**Comment by dsfrederic** [comment]
@karlschriek @nkvuong any update on this topic? 

This should have priority imho.

**Comment by orolega** [comment]
+1 to this feature.

**Comment by pedroGitt** [comment]
+1 
We are in multi project environment, and it is jut not possible to establish a reliable configuration on the unity catalog metastore. 
As explained, the buggy `resource "databricks_grants"` API is removing all the existing granting configuration. 
This defect should be in high priority.

The only workaround for now, is that **all the parties** involved in granting privileges in the metastore/catalog..  use the [Update permissions databricks API](https://docs.databricks.com/api/azure/workspace/grants/update) which has a nice `add `and `remove` API approach. 

Any update about this topic? It lasts for February..

**Comment by martin-walsh** [comment]
https://github.com/databricks/terraform-provider-databricks/pull/3024 addresses `databricks_grant` feature request. The backing API is quite nice as it allows PATCH updates for changes only.

`databricks_permission` is not so easy as the backing API is a full replacement PUT update.

**Comment by alexott** [comment]
@martin-walsh there is PATCH support for permissions as well: https://docs.databricks.com/api/workspace/tokenmanagement/updatepermissions

**Comment by martin-walsh** [comment]
> @martin-walsh there is PATCH support for permissions as well: https://docs.databricks.com/api/workspace/tokenmanagement/updatepermissions

Oh nice, and that won't overwrite all the existing permissions for the token ?

**Comment by alexott** [comment]
yes, it's in general API design: PATCH is doing partial update, and PUT is overwriting everything

**Comment by nkvuong** [comment]
@kvedes `databricks_grant` relaxes the restriction of `databricks_grants`, but it is still authoritative for a given principal, so you need to specify all the permissions for a principal in a single resource.

**Comment by kvedes** [comment]
@nkvuong Thank you, I thought I had tested that but it seems to work. 
I would suggest that the docs are updated to reflect the difference between `databricks_grant` and `databricks_grants` in relation to when they are authoritative.

**Comment by kvedes** [comment]
Is `databricks_permission` on the roadmap? - Would be really nice to have this. It is extremely difficult to make modular code when using `databricks_permissions`.

**Comment by kvedes** [comment]
@karlschriek I believe you can remove the "needs-api" label, to make it clear that this ticket is available for some one to take up.

**Comment by Tanchwa** [comment]
hey I'm working on the databricks permission resource but I hit a couple issues already

Is the doccumentation for contributing new resources not up to date anymore? I see that instead of a list of resources in `/internal/providers/pluginfw/pluginfw.go`, the Resources() factory function doesn't return an array of functions anymore but another helper function `getPluginFrameworkResourcesToRegister()` 

Are we no longer adding resources directly to that? What's the prefered method now?

**Comment by nkvuong** [comment]
hi @Tanchwa , sorry the contribution documentation is a bit out of date, as we are focusing on internal work to simplify onboarding new resources going forward for Terraform provider.

if you want to open a draft PR for some discussion/review, please go ahead and we will try to help

**Comment by Tanchwa** [comment]
Do we have a preference for how we're going to read the permissions over the API and save them to state? Right now, the GET command in the SDK/ API returns the entire set of permissions regardless of if terraform created it or not. Easiest way would probably just to chuck those into state as well. 

Otherwise we're going to have to make some decisions about what the provider should consider "changed" in order for it to properly grab the information about the permissions it knows about (I'm thinking it would need to conditionally loop through that list of objects returned).

**Comment by pcsrijith** [comment]
@nkvuong , @alexott Bringing back this old thread considering we are still waiting for this feature. Can we please get some traction on this considering it's been pending from years.

**Comment by mathiasjespers** [comment]
Is there still a track to tackle this issue? It's impossible for my company to manage cluster policy permissions using terraform because of this at the moment
